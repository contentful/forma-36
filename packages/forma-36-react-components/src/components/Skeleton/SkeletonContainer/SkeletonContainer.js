import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import styles from './SkeletonContainer.css';

export class SkeletonContainer extends React.Component {
  static propTypes = {
    extraClassNames: PropTypes.string,
    children: PropTypes.node.isRequired,
    testId: PropTypes.string,
    ariaLabel: PropTypes.string,
    width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    height: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    preserveAspectRatio: PropTypes.string,
    clipId: PropTypes.string,
    gradientId: PropTypes.string,
    backgroundColor: PropTypes.string,
    backgroundOpacity: PropTypes.number,
    animate: PropTypes.bool,
    speed: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    foregroundColor: PropTypes.string,
    foregroundOpacity: PropTypes.number,
    svgWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
    svgHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  };

  static defaultProps = {
    extraClassNames: undefined,
    testId: 'cf-ui-skeleton-form',
    ariaLabel: 'Loading component...',
    width: '100%',
    height: '100%',
    preserveAspectRatio: 'xMidYMid meet',
    clipId: 'cf-ui-skeleton-clip-id',
    gradientId: 'cf-ui-skeleton-clip-gradient',
    backgroundColor: '#e5ebed',
    backgroundOpacity: 1,
    animate: true,
    speed: 2,
    foregroundColor: '#f7f9fa',
    foregroundOpacity: 1,
    svgWidth: '100%',
    svgHeight: '100%',
  };

  render() {
    const {
      extraClassNames,
      children,
      testId,
      ariaLabel,
      width,
      height,
      preserveAspectRatio,
      clipId,
      gradientId,
      backgroundColor,
      backgroundOpacity,
      animate,
      speed,
      foregroundColor,
      foregroundOpacity,
      svgWidth,
      svgHeight,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.SkeletonContainer, extraClassNames);

    return (
      <svg
        role="img"
        className={classNames}
        aria-label={ariaLabel}
        preserveAspectRatio={preserveAspectRatio}
        width={svgWidth}
        height={svgHeight}
        {...otherProps}
      >
        {ariaLabel ? <title>{ariaLabel}</title> : null}
        <rect
          x="0"
          y="0"
          width={width}
          height={height}
          clipPath={`url(#${clipId})`}
          style={{ fill: `url(#${gradientId})` }}
        />

        <defs>
          <clipPath id={clipId}>{children}</clipPath>

          <linearGradient id={gradientId}>
            <stop
              offset="0%"
              stopColor={backgroundColor}
              stopOpacity={backgroundOpacity}
            >
              {animate && (
                <animate
                  id="animation1"
                  attributeName="stop-color"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
            <stop
              offset="50%"
              stopColor={foregroundColor}
              stopOpacity={foregroundOpacity}
            >
              {animate && (
                <animate
                  attributeName="stop-color"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  begin="animation1.begin+0.25s"
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
            <stop
              offset="100%"
              stopColor={backgroundColor}
              stopOpacity={backgroundOpacity}
            >
              {animate && (
                <animate
                  attributeName="stop-color"
                  begin="animation1.begin+0.5s"
                  values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                  dur={`${speed}s`}
                  repeatCount="indefinite"
                />
              )}
            </stop>
          </linearGradient>
        </defs>
      </svg>
    );
  }
}

export default SkeletonContainer;
