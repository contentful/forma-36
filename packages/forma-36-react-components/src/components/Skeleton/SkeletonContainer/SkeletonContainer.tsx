import React from 'react';
import cn from 'classnames';
import styles from './SkeletonContainer.css';

export interface SkeletonContainerProps {
  width?: number | string;
  height?: number | string;
  preserveAspectRatio?: string;
  clipId?: string;
  gradientId?: string;
  animateId?: string;
  backgroundColor?: string;
  backgroundOpacity?: number;
  animate?: boolean;
  speed?: number | string;
  foregroundColor?: string;
  foregroundOpacity?: number;
  svgWidth?: string | number;
  svgHeight?: string | number;

  ariaLabel?: string;
  className?: string;
  testId?: string;
  children: React.ReactNode;
}

let idCounter = 0;

export function SkeletonContainer({
  className,
  children,
  testId = 'cf-ui-skeleton-form',
  ariaLabel = 'Loading component...',
  width = '100%',
  height = '100%',
  preserveAspectRatio = 'xMidYMid meet',
  clipId = `cf-ui-skeleton-clip-${idCounter++}`,
  gradientId = `cf-ui-skeleton-clip-gradient-${idCounter++}`,
  animateId = `animation-${idCounter++}`,
  backgroundColor = '#e5ebed',
  backgroundOpacity = 1,
  animate = true,
  speed = 2,
  foregroundColor = '#f7f9fa',
  foregroundOpacity = 1,
  svgWidth = '100%',
  svgHeight = '100%',
  ...otherProps
}: SkeletonContainerProps): React.ReactElement {
  const classNames = cn(styles['SkeletonContainer'], className);

  return (
    <svg
      role="img"
      className={classNames}
      aria-label={ariaLabel}
      preserveAspectRatio={preserveAspectRatio}
      width={svgWidth}
      height={svgHeight}
      data-test-id={testId}
      {...otherProps}
    >
      {ariaLabel ? <title>{ariaLabel}</title> : null}
      <rect
        x="0"
        y="0"
        width={width}
        height={height}
        clipPath={`url(#${clipId})`}
        style={{ fill: `url(#${gradientId})` }}
      />

      <defs>
        <clipPath id={clipId}>{children}</clipPath>

        <linearGradient id={gradientId}>
          <stop
            offset="0%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {animate && (
              <animate
                id={animateId}
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="50%"
            stopColor={foregroundColor}
            stopOpacity={foregroundOpacity}
          >
            {animate && (
              <animate
                attributeName="stop-color"
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                begin={`${animateId}.begin+0.25s`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
          <stop
            offset="100%"
            stopColor={backgroundColor}
            stopOpacity={backgroundOpacity}
          >
            {animate && (
              <animate
                attributeName="stop-color"
                begin={`${animateId}.begin+0.5s`}
                values={`${backgroundColor}; ${foregroundColor}; ${backgroundColor}`}
                dur={`${speed}s`}
                repeatCount="indefinite"
              />
            )}
          </stop>
        </linearGradient>
      </defs>
    </svg>
  );
}
