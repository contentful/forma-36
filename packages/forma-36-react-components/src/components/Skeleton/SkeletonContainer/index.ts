export * from './SkeletonContainer';
export { default } from './SkeletonContainer';
