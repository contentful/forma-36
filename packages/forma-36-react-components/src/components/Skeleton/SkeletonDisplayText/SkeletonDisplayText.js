import React from 'react';
import SkeletonText from '../SkeletonText';
import { SkeletonTextPropTypes } from '../SkeletonText/SkeletonText';

export class SkeletonDisplayText extends React.Component {
  static propTypes = SkeletonTextPropTypes;

  static defaultProps = {
    numberOfLines: 1,
    width: 100,
    offsetTop: 0,
    offsetLeft: 0,
    lineHeight: 21,
    marginBottom: 20,
  };

  render() {
    return <SkeletonText {...this.props} />;
  }
}

export default SkeletonDisplayText;
