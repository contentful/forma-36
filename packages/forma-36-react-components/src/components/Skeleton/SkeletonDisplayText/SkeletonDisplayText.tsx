import React, { Component } from 'react';
import SkeletonText from '../SkeletonText';
import { SkeletonTextProps } from '../SkeletonText/SkeletonText';

export class SkeletonDisplayText extends Component<SkeletonTextProps> {
  static defaultProps = {
    numberOfLines: 1,
    width: 100,
    offsetTop: 0,
    offsetLeft: 0,
    lineHeight: 21,
    marginBottom: 20,
  };

  render() {
    return <SkeletonText {...this.props} />;
  }
}

export default SkeletonDisplayText;
