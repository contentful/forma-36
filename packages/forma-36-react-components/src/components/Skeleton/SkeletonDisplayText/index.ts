export * from './SkeletonDisplayText';
export { default } from './SkeletonDisplayText';
