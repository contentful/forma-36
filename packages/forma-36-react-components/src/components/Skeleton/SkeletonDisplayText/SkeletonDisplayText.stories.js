import React from 'react';
import { storiesOf } from '@storybook/react';
import { number } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';

import SkeletonContainer from '../SkeletonContainer';
import SkeletonDisplayText from './SkeletonDisplayText';

storiesOf('Components|Skeleton/SkeletonDisplayText', module)
  .addDecorator(withInfo)
  .add('default', () => (
    <SkeletonContainer>
      <SkeletonDisplayText numberOfLines={number('Number of lines', 1)} />
    </SkeletonContainer>
  ));
