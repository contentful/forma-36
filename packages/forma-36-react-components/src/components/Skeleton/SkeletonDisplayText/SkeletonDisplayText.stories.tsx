import React from 'react';
import { storiesOf } from '@storybook/react';
import { number } from '@storybook/addon-knobs';

import SkeletonContainer from '../SkeletonContainer';
import SkeletonDisplayText from './SkeletonDisplayText';

storiesOf('Components|Skeleton/SkeletonDisplayText', module)
  .addParameters({
    propTypes: SkeletonDisplayText['__docgenInfo'],
  })
  .add('default', () => (
    <SkeletonContainer>
      <SkeletonDisplayText numberOfLines={number('Number of lines', 1)} />
    </SkeletonContainer>
  ));
