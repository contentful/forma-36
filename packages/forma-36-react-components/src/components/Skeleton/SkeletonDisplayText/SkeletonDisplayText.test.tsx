import React from 'react';
import { shallow } from 'enzyme';
import axe from '../../../utils/axeHelper';
import SkeletonContainer from '../SkeletonContainer';
import SkeletonDisplayText from './SkeletonDisplayText';

it('renders the component', () => {
  const output = shallow(
    <SkeletonContainer>
      <SkeletonDisplayText />
    </SkeletonContainer>,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = shallow(
    <SkeletonContainer>
      <SkeletonDisplayText />
    </SkeletonContainer>,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
