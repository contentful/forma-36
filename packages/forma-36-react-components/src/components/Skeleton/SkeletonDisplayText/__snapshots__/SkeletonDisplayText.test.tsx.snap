// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<svg
  aria-label="Loading component..."
  className="SkeletonContainer"
  height="100%"
  preserveAspectRatio="xMidYMid meet"
  role="img"
  width="100%"
>
  <title>
    Loading component...
  </title>
  <rect
    clipPath="url(#cf-ui-skeleton-clip-id)"
    height="100%"
    style={
      Object {
        "fill": "url(#cf-ui-skeleton-clip-gradient)",
      }
    }
    width="100%"
    x="0"
    y="0"
  />
  <defs>
    <clipPath
      id="cf-ui-skeleton-clip-id"
    >
      <SkeletonDisplayText
        lineHeight={21}
        marginBottom={20}
        numberOfLines={1}
        offsetLeft={0}
        offsetTop={0}
        width={100}
      />
    </clipPath>
    <linearGradient
      id="cf-ui-skeleton-clip-gradient"
    >
      <stop
        offset="0%"
        stopColor="#e5ebed"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          dur="2s"
          id="animation1"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="50%"
        stopColor="#f7f9fa"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          begin="animation1.begin+0.25s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
      <stop
        offset="100%"
        stopColor="#e5ebed"
        stopOpacity={1}
      >
        <animate
          attributeName="stop-color"
          begin="animation1.begin+0.5s"
          dur="2s"
          repeatCount="indefinite"
          values="#e5ebed; #f7f9fa; #e5ebed"
        />
      </stop>
    </linearGradient>
  </defs>
</svg>
`;
